src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrUpdater.cs

```diff
@@ -226,33 +231,122 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
-            currentVersion.Sha,
-            update.TargetRevision);
-
-        return await UpdateRepoToRevisionAsync(
+        await StageRepositoryUpdatesAsync(
             update,
             clone,
             additionalRemotes,
             currentVersion.Sha,
-            author: null,
-            commitMessage,
-            reapplyVmrPatches,
             componentTemplatePath,
             tpnTemplatePath,
             generateCodeowners,
             discardPatches,
             cancellationToken);
+
+        if (reapplyVmrPatches)
+        {
+            await ReapplyVmrPatchesAsync(cancellationToken);
+        }
+
+        var commitMessage = PrepareCommitMessage(
+            SquashCommitMessage,
+            update.Mapping.Name,
+            update.RemoteUri,
+            currentVersion.Sha,
+            update.TargetRevision);
+
+        await CommitAsync(commitMessage);
+
+        return true;
+    }
+    /// <summary>
+    /// Updates a repository and all of it's dependencies recursively starting with a given mapping.
+    /// Always updates to the first version found per repository in the dependency tree.
+    /// </summary>
+    private async Task<bool> UpdateRepositoriesRecursively(
+        VmrDependencyUpdate rootUpdate,
+        IReadOnlyCollection<AdditionalRemote> additionalRemotes,
+        string? componentTemplatePath,
+        string? tpnTemplatePath,
+        bool generateCodeowners,
+        bool discardPatches,
+        CancellationToken cancellationToken)
+    {
+        // Synchronization creates commits (one per mapping and some extra)
+        // on a separate branch that is then merged into original one
+        var originalRootSha = Commit.GetShortSha(GetCurrentVersion(rootUpdate.Mapping));
+        var workBranchName = "sync" +
+            $"/{rootUpdate.Mapping.Name}" +
+            $"/{originalRootSha}-{rootUpdate.TargetRevision}";
+        IWorkBranch workBranch = await _workBranchFactory.CreateWorkBranchAsync(LocalVmr, workBranchName);
+
+        await RestoreVmrPatchedFilesAsync(cancellationToken);
+        await CommitAsync("[VMR patches] Removed VMR patches");
+
+        HashSet<VmrDependencyUpdate> updatedDependencies;
+        try
+        {
+            updatedDependencies = await UpdateRepositoryRecursively(
+                rootUpdate,
+                additionalRemotes,
+                componentTemplatePath,
+                tpnTemplatePath,
+                generateCodeowners,
+                discardPatches,
+                cancellationToken);
+        }
+        catch (Exception)
+        {
+            _logger.LogWarning(
+                InterruptedSyncExceptionMessage,
+                workBranch.OriginalBranch.StartsWith("sync") || workBranch.OriginalBranch.StartsWith("init")
+                    ? "the original"
+                    : workBranch.OriginalBranch);
+            throw;
+        }
+
+        await ReapplyVmrPatchesAsync(cancellationToken);
```

Comment:
Might be confusing things here, but why are we applying the path twice? Once before the sync (and we remove the patch files themselves) and then afterwards here where we keep the patches.

Reply:
What do you mean twice?

Before the sync, we remove all VMR patches from the VMR so that the content matches what is in the repos and then after we bring in the changes from the repos, we re-apply the VMR patches back.

Reply:
Oh, I looked into `RestoreVmrPatchedFileAsync` and saw that we are calling the same `ApplyPatch` method as in `ReapplyVmrPatchesAsync`, but failed to notice that `applyReverse` is set. Guessing that is the difference between the two